File: AdventOfCode2023/Days/Day2.cs
namespace AdventOfCode2023.Solutions;

public record Day2 : Day<Day2>, IDay<Day2>
{
    public static int DayNumber => 2;

    public static object SolvePart1(ImmutableArray<string> input) => input.Select(Parse).Where(Valid).Sum(_ => _.Id);

    public static object SolvePart2(ImmutableArray<string> input) => input.Select(Parse).Sum(Power);

    public static Game Parse(string line)
    {
        static Dictionary<string, int> ParseRound(string roundString) => roundString
            .Split(',')
            .Select(_ => _.Trim().Split(' '))
            .ToDictionary(_ => _[1], _ => int.Parse(_[0]));

        var parts = line.Split(':');

        return new(
            Id: int.Parse(parts[0][5..]),
            Rounds: parts[1].Split(';').Select(ParseRound).ToArray()
        );
    }

    public static bool Valid(Game game) =>
        game.Rounds.Max(_ => _.ValueOrZero("red")) <= 12
        && game.Rounds.Max(_ => _.ValueOrZero("green")) <= 13
        && game.Rounds.Max(_ => _.ValueOrZero("blue")) <= 14;

    public static int Power(Game game) =>
        game.Rounds.Max(_ => _.ValueOrZero("red"))
        * game.Rounds.Max(_ => _.ValueOrZero("green"))
        * game.Rounds.Max(_ => _.ValueOrZero("blue"));

    public record Game(int Id, Dictionary<string, int>[] Rounds);
}

File: AdventOfCode2023.FSharp/Days/Day2.fs
module Day2
    open System.Collections.Generic
    open System.IO
    open Utility

    type Game = { Id: int; Rounds : IDictionary<string, int>[] } with
        member this.MaxBallsByColour colour =
            this.Rounds |> Seq.map (valueOrDefault colour 0) |> Seq.max

    let parseRound roundText =
        roundText
        |> split ','
        |> Seq.map (fun x -> x.Trim().Split(' '))
        |> Seq.map (fun x -> (x[1], int(x[0])))
        |> dict

    let parseLine line =
        line
        |> split ':'
        |> Seq.item 1
        |> split ';'
        |> Seq.map parseRound
        |> Seq.toArray

    let parseInput() =
        "Input/day2.txt"
        |> File.ReadLines
        |> Seq.map parseLine
        |> Seq.indexed
        |> Seq.map (fun x -> {Id = fst(x) + 1; Rounds = snd(x)})

    let validateGame (game : Game) =
        game.MaxBallsByColour "red" < 13
        && game.MaxBallsByColour "green" < 14
        && game.MaxBallsByColour "blue" < 15

    let getPower (game : Game) =
        game.MaxBallsByColour "red"
        * game.MaxBallsByColour "green"
        * game.MaxBallsByColour "blue"

    let solvePart1() =
        parseInput()
        |> Seq.filter validateGame
        |> Seq.map (fun x -> x.Id)
        |> Seq.sum

    let solvePart2() =
        parseInput()
        |> Seq.map getPower
        |> Seq.sum